I have unapplied the patch in this MR and updated the test as I suggested above. Running this test without the containerd storage driver, I can demonstrate the updated test catches the bug described by #51364. That is, the test identifies the contemporary timestamp on the `blobs/{digest}` directory is newer than `img.Created`.

```
$ DOCKER_GRAPHDRIVER=overlay2 TEST_INTEGRATION_USE_GRAPHDRIVER="1" TESTFLAGS="-test.run ^TestSaveCheckTimes$" make test-integration
...
Running /go/src/github.com/docker/docker/integration/image (amd64.docker.docker.integration.image) flags=-test.v -test.timeout=10m -test.run ^TestSaveCheckTimes$                                                      12:30:03 [184/1888]
INFO: Testing against a local daemon
=== RUN   TestSaveCheckTimes
=== PAUSE TestSaveCheckTimes
=== CONT  TestSaveCheckTimes
    save_test.go:86: checking blobs/ (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/ (modtime=2025-11-03 12:30:03 +0000 UTC)
    save_test.go:87: assertion failed: modtime.After(threshold) is true: blobs/sha256/ has modtime 2025-11-03 12:30:03 +0000 UTC after 2020-06-02 21:19:57.279412246 +0000 UTC
    save_test.go:86: checking blobs/sha256/1be74353c3d0fd55fb5638a52953e6f1bc441e5b1710921db9ec2aa202725569 (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/1c35c441208254cb7c3844ba95a96485388cef9ccc0646d562c7fc026e04c807 (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/5d68854a5204233cd7872a5a6346afcca085cd5af4d3be7c3d26f795248350bb (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/5fd46bbd12cd97c83beb8a563186270bb09d0e62ad43c024859c1c2bfb20fb0f (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking index.json (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking manifest.json (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking oci-layout (modtime=1970-01-01 00:00:00 +0000 UTC)                                                                                                                                                           
    save_test.go:86: checking repositories (modtime=1970-01-01 00:00:00 +0000 UTC)
--- FAIL: TestSaveCheckTimes (0.07s)
FAIL
```

The containerd case passes as expected as all the timestamps appear to be sanitised to mtime 0 anyway:

```
$ TESTFLAGS="-test.run ^TestSaveCheckTimes$" make test-integration
...
Running /go/src/github.com/docker/docker/integration/image (amd64.docker.docker.integration.image) flags=-test.v -test.timeout=10m -test.run ^TestSaveCheckTimes$
INFO: Testing against a local daemon                                                                                                                                                                                                      === RUN   TestSaveCheckTimes
=== PAUSE TestSaveCheckTimes                                                                                                                                                                                                              === CONT  TestSaveCheckTimes
    save_test.go:86: checking blobs/ (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/ (modtime=1970-01-01 00:00:00 +0000 UTC)                                                                                                                                                           
    save_test.go:86: checking blobs/sha256/1c35c441208254cb7c3844ba95a96485388cef9ccc0646d562c7fc026e04c807 (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/328dac05c07eb589d3a783c8bd2bc264f38dbcf1f71d00844804bd4c4aba3597 (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/76df9210b28cbd4bc127844914d0a23937ed213048dc6289b2a2d4f7d675c75e (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking index.json (modtime=1970-01-01 00:00:00 +0000 UTC)                                                                                                                                                              
    save_test.go:86: checking manifest.json (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking oci-layout (modtime=1970-01-01 00:00:00 +0000 UTC)
--- PASS: TestSaveCheckTimes (0.20s)
PASS
```

Reapplying the patch in this MR, the test now passes under the overlay2 condition as the `blobs/{digest}` directory is now set to 0. In retrospect, perhaps this should actually be set to img.Created...

```
=== RUN   TestSaveCheckTimes
=== PAUSE TestSaveCheckTimes                                                                                                                                                                                                              === CONT  TestSaveCheckTimes
    save_test.go:86: checking blobs/ (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/ (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/1be74353c3d0fd55fb5638a52953e6f1bc441e5b1710921db9ec2aa202725569 (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/1c35c441208254cb7c3844ba95a96485388cef9ccc0646d562c7fc026e04c807 (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking blobs/sha256/5d68854a5204233cd7872a5a6346afcca085cd5af4d3be7c3d26f795248350bb (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking blobs/sha256/5fd46bbd12cd97c83beb8a563186270bb09d0e62ad43c024859c1c2bfb20fb0f (modtime=2020-06-02 21:19:57 +0000 UTC)
    save_test.go:86: checking index.json (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking manifest.json (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking oci-layout (modtime=1970-01-01 00:00:00 +0000 UTC)
    save_test.go:86: checking repositories (modtime=1970-01-01 00:00:00 +0000 UTC)
--- PASS: TestSaveCheckTimes (0.05s)                                                                                                                                                                                                      PASS
```

I've updated the commit in this MR to include the changes to the test. Note that I temporarily added a log line to the test to indicate the tar member name and mtime, just to show to you here that all tar members are checked and to emphasise the existing difference in behaviour between the two storage drivers; this log line is not in the MR.

I believe this MR now correctly tests for and fixes the unexpected behaviour described by #51364. I'd suggest further discussion on whether the mtimes should always be 0 is a new issue, as the existing code clearly indicates an intention to maintain img.Created on some of the tar members, for reasons unknown to me. I'd be happy to make this change and submit a new MR to change mtime to 0 in all cases if that is desired.

--- a/daemon/internal/image/tarexport/save.go
+++ b/daemon/internal/image/tarexport/save.go
@@ -264,14 +264,14 @@ func (s *saveSession) save(ctx context.Context, outStream io.Writer) error {
 		if err := mkdirAllWithChtimes(filepath.Dir(mFile), 0o755, time.Unix(0, 0), time.Unix(0, 0)); err != nil {
 			return errors.Wrap(err, "error creating blob directory")
 		}
-		if err := system.Chtimes(filepath.Dir(mFile), time.Unix(0, 0), time.Unix(0, 0)); err != nil {
-			return errors.Wrap(err, "error setting blob directory timestamps")
-		}
 		if err := os.WriteFile(mFile, data, 0o644); err != nil {
 			return errors.Wrap(err, "error writing oci manifest file")
 		}
 		if err := system.Chtimes(mFile, time.Unix(0, 0), time.Unix(0, 0)); err != nil {
-			return errors.Wrap(err, "error setting blob directory timestamps")
+			return errors.Wrap(err, "error setting oci manifest timestamp")
+		}
+		if err := system.Chtimes(filepath.Dir(mFile), time.Unix(0, 0), time.Unix(0, 0)); err != nil {
+			return errors.Wrap(err, "error setting blob digest directory timestamp")
 		}
 
 		untaggedMfstDesc := ocispec.Descriptor{

--- a/integration/image/save_test.go
+++ b/integration/image/save_test.go
@@ -65,26 +65,25 @@ func TestSaveCheckTimes(t *testing.T) {
 	rdr, err := apiClient.ImageSave(ctx, []string{repoName})
 	assert.NilError(t, err)
 
-	tarfs := tarIndexFS(t, rdr)
-
-	dt, err := fs.ReadFile(tarfs, "manifest.json")
-	assert.NilError(t, err)
-
-	var ls []imageSaveManifestEntry
-	assert.NilError(t, json.Unmarshal(dt, &ls))
-	assert.Assert(t, is.Len(ls, 1))
-
-	info, err := fs.Stat(tarfs, ls[0].Config)
-	assert.NilError(t, err)
-
 	created, err := time.Parse(time.RFC3339, img.Created)
 	assert.NilError(t, err)
 
-	if testEnv.UsingSnapshotter() {
-		// containerd archive export sets the mod time to zero.
-		assert.Check(t, is.Equal(info.ModTime(), time.Unix(0, 0)))
-	} else {
-		assert.Check(t, is.Equal(info.ModTime().Format(time.RFC3339), created.Format(time.RFC3339)))
+	// containerd archive export sets mod times of all members to zero
+	// otherwise no member should be newer than the image created date
+	threshold := time.Unix(0, 0)
+	if !testEnv.UsingSnapshotter() {
+		threshold = created
+	}
+
+	tr := tar.NewReader(rdr)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		assert.NilError(t, err)
+		modtime := hdr.ModTime
+		assert.Check(t, !modtime.After(threshold), "%s has modtime %s after %s", hdr.Name, modtime, threshold)
 	}
 }
 

